components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,381 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+use std::rc::Rc;
+
+use base::id::GamepadRouterId;
+use dom_struct::dom_struct;
+use embedder_traits::{DualRumbleEffectParams, EmbedderMsg};
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+use js::jsval::JSVal;
+use script_traits::{GamepadSupportedHapticEffects, ScriptMsg};
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::GamepadHapticActuatorBinding::{
+    GamepadEffectParameters, GamepadHapticActuatorMethods, GamepadHapticEffectType,
+};
+use crate::dom::bindings::codegen::Bindings::PerformanceBinding::Performance_Binding::PerformanceMethods;
+use crate::dom::bindings::codegen::Bindings::WindowBinding::Window_Binding::WindowMethods;
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::refcounted::{Trusted, TrustedPromise};
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::bindings::utils::to_frozen_array;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::realms::{AlreadyInRealm, InRealm};
+use crate::script_runtime::JSContext;
+use crate::task::TaskCanceller;
+use crate::task_source::gamepad::GamepadTaskSource;
+use crate::task_source::{TaskSource, TaskSourceName};
+
+struct HapticEffectStopListener {
+    canceller: TaskCanceller,
+    task_source: GamepadTaskSource,
+    context: Trusted<GamepadHapticActuator>,
+}
+
+impl HapticEffectStopListener {
+    fn handle(&self, stopped_successfully: bool) {
+        let context = self.context.clone();
+
+        let _ = self.task_source.queue_with_canceller(
+            task!(handle_haptic_effect_stopped: move || {
+                let actuator = context.root();
+                actuator.handle_haptic_effect_stopped(stopped_successfully);
+            }),
+            &self.canceller,
+        );
+    }
+}
+
+#[dom_struct]
+pub struct GamepadHapticActuator {
+    reflector_: Reflector,
+    gamepad_index: u32,
+    effects: Vec<GamepadHapticEffectType>,
+    #[ignore_malloc_size_of = "Rc is hard"]
+    playing_effect_promise: DomRefCell<Option<Rc<Promise>>>,
+    sequence_id: Cell<u32>,
```

Comment:
Let's add some doc explaining what they are used for(since the spec isn't updated yet you could link to the PR).